allennlp/data/data_loaders/multitask_data_loader.py

```diff
@@ -0,0 +1,264 @@
+        instance_queue_size: Dict[str, int] = None,
+        instance_chunk_size: Dict[str, int] = None,
+    ) -> None:
+        self.readers = reader.readers
+        self.data_paths = data_path
+        self.scheduler = scheduler
+        self.sampler = sampler
+
+        self._batch_size = batch_size
+        self._instances_per_epoch = instances_per_epoch
+        self._batch_size_multiplier = batch_size_multiplier or {}
+        for multiplier in self._batch_size_multiplier.values():
+            if multiplier > batch_size:
+                raise ValueError(
+                    f"Multiplier value ({multiplier}) is larger than batch size ({batch_size})"
+                )
+        self._drop_last = drop_last
+
+        self._num_workers = num_workers or {}
+        self._max_instances_in_memory = max_instances_in_memory or {}
+        self._start_method = start_method or {}
+        self._instance_queue_size = instance_queue_size or {}
+        self._instance_chunk_size = instance_chunk_size or {}
+
+        if self.readers.keys() != self.data_paths.keys():
+            raise ValueError(
+                f"Mismatch between readers ({self.readers.keys()}) and data paths"
+                " ({self.data_paths.keys()})"
+            )
+        self._loaders = {key: self._make_data_loader(key) for key in self.readers}
+
+        # This stores our current iterator with each dataset, so we don't just iterate over the
+        # first k instances every epoch if we're using instances_per_epoch.  We'll grab instances
+        # from here each epoch, and refresh it when it runs out.  We only use this in the case that
+        # batchs_per_epoch is not None, but these iterators are lazy, so always creating them
+        # doesn't hurt anything.
+        self._iterators: Dict[str, Iterator[Instance]] = {
+            # NOTE: The order in which we're calling these iterator functions is important.  We want
+            # an infinite iterator over the data, but we want the order in which we iterate over the
+            # data to be different at every epoch.  The cycle function will give us an infinite
+            # iterator, and it will call the lambda function each time it runs out of instances,
+            # which will produce a new shuffling of the dataset.
+            key: util.cycle_iterator_function(
+                # This default argument to the lambda function is necessary to create a new scope
+                # for the loader variable, so a _different_ loader gets saved for every iterator.
+                # Dictionary comprehensions don't create new scopes in python.  If you don't have
+                # this loader, you end up with `loader` always referring to the last loader in the
+                # iteration...  mypy also doesn't know what to do with this, for some reason I can't
+                # figure out.
+                lambda l=loader: util.shuffle_iterable(l.iter_instances())  # type: ignore
```

Comment:
Everywhere else we make shuffling optional, because when you're debugging you want the same order every time.